keps/sig-network/20200204-NetworkPolicy-verification-rearchitecture.md

```diff
@@ -0,0 +1,668 @@
+ 
+tests.  These tests use almost identical harnesses, with a subtle `},{` clause
+differentiating the stacked peers in network policy (an *or* selector) vs. a combined peer policy.
+ 
+```
+ 
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                // TODO add these composably, so that can be disambiguated from combo networkpolicypeer
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                            },
+                            {
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+                    }},
+```
+ 
+The AND test is obviously more selective, although it is tricky to tell from the struct that
+it has been correctly written to be different from the OR test.
+ 
+```
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                                // because we lack },{ , this is a single Peer selecting
+                                // pods, from namespaces selected by namespace selector.
+                                // This is difficult to verify for correctness at a
+                                // glance, due to the verbosity of the struct.
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+```
+
+In contrast, we can express the same and test using the API (including its entire connectivity matrix) in this proposal as follows:
+
+```
+	builder := &NetworkPolicySpecBuilder{}
+	builder = builder.SetName("myns", "allow-podb-in-nsb").SetPodSelector(map[string]string{"pod": "b"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod-name": "b"}, nil, map[string]string{"ns-name": "b"}, nil)
```

Comment:
I continue to strongly reject the idea that this is a more obvious representation of the policy than the explicit `struct` form. What do all of those different `nil`s mean? This just requires people to learn a *different* representation, which is not used anywhere except in these test cases, which thus makes it *more* difficult for people to contribute to them.

Reply:
Hmm, havent heard this criticism before.  Is this a blocker for you?

Reply:
I said it before, https://github.com/kubernetes/enhancements/pull/1568#discussion_r397333193

I wouldn't 100% block the KEP based on this but I'd like to hear what other people think.

Reply:
@danwinship  , either way,  this is doable, so since its not hard to make it more descriptive, I added this to the KEP:
```	// Note: We also can implement
	// a builder that is more descriptive, such as, 
	// builder.AddIngress().WithProtocol(tcp).ToPort(80).FromNs("x"), in the final implementation...
```
Would you prefer we do that ? Im ok either way.

Reply:
I haven't been involved with sig-network before, but I am working with Jay in this space. I've had good experience with a builder helper to reduce the amount of definitions you have to spell out as full go structs in test code. I generally find that it's much cleaner assuming the builders are clean and fairly compact (see, for example https://github.com/vmware-tanzu/velero/tree/master/pkg/builder).

I do, however, agree with you @danwinship that it's not intuitive when you have builder functions that take in 8 parameters and a lot of them end up being nil. Something like AddIngress could probably be broken down into further builder functions.

Reply:
I think sounds like we are in agreement on both sides - we can  agree to make a more granular DSL here for the final impl.

Reply:
The usual golang solution is to use a struct so you can have named fields:

AddIngress(&IngressBuilderParams{field1: , field2: , field3: }

Reply:
thats a reasonable third option, ill add that to my comment below.

Reply:
marking as resolved, see the comments above the builder which demark it as *downvoted* in favor of structs.